core/trino-spi/src/main/java/io/trino/spi/predicate/SortedRangeSet.java

```diff
@@ -48,54 +73,162 @@
+            if (sortedRanges.isNull(position)) {
+                if (inclusive[position]) {
+                    throw new IllegalArgumentException("Invalid inclusive marker for null value at position " + position);
+                }
+                if (position != 0 && position != sortedRanges.getPositionCount() - 1) {
+                    throw new IllegalArgumentException(format("Invalid null value at position %s of %s", position, sortedRanges.getPositionCount()));
+                }
+            }
+        }
+        this.inclusive = inclusive;
+        this.sortedRanges = sortedRanges;
     }
 
     static SortedRangeSet none(Type type)
     {
-        return copyOf(type, Collections.emptyList());
+        return new SortedRangeSet(
+                type,
+                new boolean[0],
+                // TODO This can perhaps use an empty block singleton
+                type.createBlockBuilder(null, 0).build());
     }
 
     static SortedRangeSet all(Type type)
     {
-        return copyOf(type, Collections.singletonList(Range.all(type)));
+        return new SortedRangeSet(
+                type,
+                new boolean[] {false, false},
+                // TODO This can perhaps use a "block with two nulls" singleton
+                type.createBlockBuilder(null, 2)
+                        .appendNull()
+                        .appendNull()
+                        .build());
+    }
+
+    @Deprecated // For JSON deserialization only
+    @JsonCreator
+    public static SortedRangeSet fromJson(
+            @JsonProperty("type") Type type,
+            @JsonProperty("inclusive") boolean[] inclusive,
+            @JsonProperty("sortedRanges") Block sortedRanges)
+    {
+        if (sortedRanges instanceof BlockBuilder) {
+            throw new IllegalArgumentException("sortedRanges must be a block: " + sortedRanges);
+        }
+        return new SortedRangeSet(type, inclusive.clone(), sortedRanges);
     }
 
     /**
      * Provided discrete values that are unioned together to form the SortedRangeSet
      */
     static SortedRangeSet of(Type type, Object first, Object... rest)
     {
-        List<Range> ranges = new ArrayList<>(rest.length + 1);
-        ranges.add(Range.equal(type, first));
+        if (rest.length == 0) {
+            return of(type, first);
+        }
+
+        MethodHandle comparisonOperator = TUPLE_DOMAIN_TYPE_OPERATORS.getComparisonOperator(type, simpleConvention(FAIL_ON_NULL, BLOCK_POSITION, BLOCK_POSITION));
+
+        BlockBuilder blockBuilder = type.createBlockBuilder(null, 1 + rest.length);
+        writeNativeValue(type, blockBuilder, first);
         for (Object value : rest) {
-            ranges.add(Range.equal(type, value));
+            writeNativeValue(type, blockBuilder, value);
         }
-        return copyOf(type, ranges);
+        Block block = blockBuilder.build();
+
+        List<Integer> indexes = new ArrayList<>(block.getPositionCount());
```

Comment:
Consider using FastUtil.IntList here. It supports sorting ints with a comparator without boxing.

Reply:
in SPI i cannot use it without copying, which i would prefer to avoid unless there is good motivation